File: afterfall-python/afterfall/noise.py
"""Seeded, vectorised gradient noise."""
import numpy as np


class Noise:
    def __init__(self, seed: int):
        rng = np.random.default_rng(seed)
        perm = rng.permutation(256)
        self.p = np.concatenate([perm, perm]).astype(np.int64)
        angles = rng.random(256) * 2 * np.pi
        self.gx, self.gy = np.cos(angles), np.sin(angles)

    def noise(self, x, y):
        x = np.asarray(x, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64)
        xi = np.floor(x).astype(np.int64)
        yi = np.floor(y).astype(np.int64)
        xf, yf = x - xi, y - yi
        xi &= 255
        yi &= 255
        u = xf * xf * xf * (xf * (xf * 6 - 15) + 10)
        v = yf * yf * yf * (yf * (yf * 6 - 15) + 10)
        p = self.p

        def grad(ix, iy, dx, dy):
            h = p[p[ix] + iy]
            return self.gx[h] * dx + self.gy[h] * dy

        n00 = grad(xi, yi, xf, yf)
        n10 = grad(xi + 1, yi, xf - 1, yf)
        n01 = grad(xi, yi + 1, xf, yf - 1)
        n11 = grad(xi + 1, yi + 1, xf - 1, yf - 1)
        a = n00 + u * (n10 - n00)
        b = n01 + u * (n11 - n01)
        return (a + v * (b - a)) * 1.41

    def fbm(self, x, y, octaves=4):
        total, amp, freq, norm = 0.0, 1.0, 1.0, 0.0
        x = np.asarray(x, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64)
        for i in range(octaves):
            total = total + amp * self.noise(x * freq + 17.31 * i, y * freq - 9.17 * i)
            norm += amp
            amp *= 0.5
            freq *= 2.03
        return total / norm


def smooth(e0, e1, x):
    """Smoothstep that also works with e0 > e1 (falls instead of rises)."""
    t = np.clip((np.asarray(x, dtype=np.float64) - e0) / (e1 - e0), 0.0, 1.0)
    return t * t * (3 - 2 * t)

File: afterfall-unity/Assets/Afterfall/Sim/Noise.cs
using System;

namespace Afterfall.Sim
{
    /// <summary>Seeded 2D gradient noise and fractal sums.</summary>
    public sealed class Noise
    {
        readonly int[] p = new int[512];
        readonly double[] gx = new double[256], gy = new double[256];

        public Noise(long seed)
        {
            var rng = new Rng(seed * 7919 + 17);
            var perm = new int[256];
            for (int i = 0; i < 256; i++) perm[i] = i;
            rng.Shuffle(perm);
            for (int i = 0; i < 512; i++) p[i] = perm[i & 255];
            for (int i = 0; i < 256; i++)
            {
                double a = rng.Random() * M.TAU;
                gx[i] = Math.Cos(a);
                gy[i] = Math.Sin(a);
            }
        }

        public double Value(double x, double y)
        {
            double fx = Math.Floor(x), fy = Math.Floor(y);
            int xi = (int)((long)fx & 255), yi = (int)((long)fy & 255);
            double xf = x - fx, yf = y - fy;
            double u = xf * xf * xf * (xf * (xf * 6 - 15) + 10);
            double v = yf * yf * yf * (yf * (yf * 6 - 15) + 10);
            double n00 = Grad(xi, yi, xf, yf), n10 = Grad(xi + 1, yi, xf - 1, yf);
            double n01 = Grad(xi, yi + 1, xf, yf - 1), n11 = Grad(xi + 1, yi + 1, xf - 1, yf - 1);
            double a = n00 + u * (n10 - n00), b = n01 + u * (n11 - n01);
            return (a + v * (b - a)) * 1.41;
        }

        double Grad(int ix, int iy, double dx, double dy)
        {
            int h = p[p[ix] + iy];
            return gx[h] * dx + gy[h] * dy;
        }

        public double Fbm(double x, double y, int octaves = 4)
        {
            double total = 0, amp = 1, freq = 1, norm = 0;
            for (int i = 0; i < octaves; i++)
            {
                total += amp * Value(x * freq + 17.31 * i, y * freq - 9.17 * i);
                norm += amp;
                amp *= .5;
                freq *= 2.03;
            }
            return total / norm;
        }
    }
}
